### PR TITLE
EDSC-3797: Added an option for concatenation for custom downloads via Harmony

### PR DESCRIPTION
**Pull request by bnp26**
# Overview
### What is the feature?
Added an option for concatenation for custom downloads via Harmony
Please summarize the feature or fix.
Added checkbox for concatenation option as well as setting the checkbox to be checked if defaultConcatenation is true in the cmr request.

### What is the Solution?

- Added a new accessMethod options associated with the Harmony service.
  - supportsConcatenation (whether the option should show up)
  - defaultConcatenation (whether the option should be checked by default)
- Implemented the new checkbox in AccessMethod
- Implemented changes in the constructOrderPayload so concatenate is set to true if enableConcatenateDownload is true.

### What areas of the application does this impact?

List impacted areas.

# Testing

### Reproduction steps

- **Environment for testing: UAT**
- **Collection to test with: C1245618475-EEDTEST** 

1. Navigate to (http://localhost:8080/search?ee=uat)
2. Put in C1245618475-EEDTEST as the collection in the search box and search.
3. Click on the collection and then select 6 granules before clicking download.
4. Click on the 2nd radio button in the data access methods list
5. Observe the combine data customization option come up with default = false so the combine data option should be off.
6. click on the checkbox and then click download.
7. Observe the download going through successfully.

### Attachments
Screenshot 2023-11-09 at 1 22 25 PM

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by trevorlang** [comment]
Do you plan on adding reproduction steps? What collection did you use for testing? I'd like to take a look locally at some point today.

**Comment by trevorlang** [comment]
Do you plan on adding reproduction steps? What collection did you use for testing? I'd like to take a look locally at some point today.

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 1]
```diff
@@ -182,48 +187,46 @@ describe('buildAccessMethods', () => {
...
-          nativeId: 'mmt_variable_3969',
-          scienceKeywords: null
-        }]
       }
     }
     const isOpenSearch = false
 
     const methods = buildAccessMethods(collectionMetadata, isOpenSearch)
 
+    console.log(methods)
```
Comment (macrouch) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:228]: Remove console log

File: static/src/js/util/accessMethods/defaultConcatenation.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  const { serviceOptions = {} } = service
+
+  // If there are no service options the record can not support variable subsetting
+  if (serviceOptions == null) return false
+
+  const { aggregation = {} } = serviceOptions
+  const { concatenate = {} } = aggregation
+
+  const { concatenateDefault = false } = concatenate
+  console.log(concatenateDefault)
```
Comment (macrouch) [inline static/src/js/util/accessMethods/defaultConcatenation.js:15]: Remove console log

File: static/src/js/util/accessMethods/supportsConcatenation.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Determine whether or not the provided UMM S record supports shapefile subsetting
+ * @param {Object} service UMM S record to parse
+ */
+export const supportsConcatenation = (service) => {
+  const { serviceOptions = {} } = service
+
+  console.log(service)
```
Comment (macrouch) [inline static/src/js/util/accessMethods/supportsConcatenation.js:8]: Remove console log

File: static/src/js/util/accessMethods/buildAccessMethods.js [hunk 4]
```diff
@@ -26,6 +28,8 @@ export const buildAccessMethods = (collectionMetadata, isOpenSearch) => {
   let associatedVariables = collectionAssociatedVariables
   const { items: serviceItems = null } = services
 
+  console.log(serviceItems)
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods.js:31]: Remove console log

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 5]
```diff
@@ -496,6 +525,38 @@ export class AccessMethod extends Component {
...
+                        <Form.Check
+                          id="input__concatinate-subsetting"
+                          type="checkbox"
+                          label={
+                            (
+                              <div>
+                                <span className="mb-1 d-block">
+                                  Enable Concatenation
+                                </span>
+                                <span className="mb-1 d-block gray-text">
```
Comment (macrouch) [inline static/src/js/components/AccessMethod/AccessMethod.js:546]: Other labels in this file are using the `text-muted` class, would that do something similar to what your `gray-text` class is doing?

If not, and you do need this class, `color: #a0a0a0;` is not the right way to define a color, we have sass variables for all the colors we should be using so we don't have hex codes all over the code

**Review by macrouch — COMMENTED** [review]
Agree with Trevor, reproduction steps need to be added. Those are also useful when you end up writing the test session

File: serverless/src/submitHarmonyOrder/constructOrderPayload.js [hunk 6]
```diff
@@ -215,6 +217,11 @@ export const constructOrderPayload = async ({
     orderPayload.append('outputCrs', selectedOutputProjection)
   }
 
+  // Adds supportsConcatenation to the payload and it's value
+  if (supportsConcatenation && enableConcatenateDownload) {
```
Comment (eudoroolivares2016) [inline serverless/src/submitHarmonyOrder/constructOrderPayload.js:221]: I was having some issues running this branch locally and a job to harmony because `enableConcatenateDownload` was always undefined. I think there are a few places where that needs to be initialized. Its not being saved in the Database right now

**Review by eudoroolivares2016 — CHANGES_REQUESTED** [review]

File: serverless/src/submitHarmonyOrder/constructOrderPayload.js (hunk 6 above)
Comment (bnp26) [reply serverless/src/submitHarmonyOrder/constructOrderPayload.js:221]: The change has been pushed up

File: static/src/js/util/accessMethods/__tests__/supportsConcatenation.test.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      })
+
+      expect(response).toBeTruthy()
+    })
+
+    test('returns true when concatenateDefault is true', () => {
+      const response = supportsConcatenation({
+        conceptId: 'S100000-EDSC',
+        serviceOptions: {
+          aggregation: {
```
Comment (bnp26) [inline static/src/js/util/accessMethods/__tests__/supportsConcatenation.test.js:24]: Add a test where aggregation exists but concatenate is not there.

**Comment by daniel-zamora** [comment]
approved per changes mentioned in the PR meeting

**Review by macrouch — DISMISSED** [review]
When checking the Enable Concatenation checkbox and refreshing the page, the form is not remembering the selected checkbox like it does with the other form fields. You need to be sure to put the value into the URL and populate it from the URL.

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js [hunk 8]
```diff
@@ -693,4 +693,115 @@ describe('constructOrderPayload', () => {
       })
     })
   })
+
+  describe('when only sending concatenate is true', () => {
+    test('constructs a payload containing supportsConcatenation = true and enableConcatenateDownload = true', async () => {
```
Comment (macrouch) [inline serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:698]: These test names are hard to follow, the payload does not contain supportsConcatenation or enableConcatenateDownload, it just contains concatenate=true

File: static/src/js/util/accessMethods/defaultConcatenation.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Determine whether or not the provided UMM S record supports shapefile subsetting
+ * @param {Object} service UMM S record to parse
+ */
+export const defaultConcatenation = (service) => {
+  const { serviceOptions = {} } = service
+
+  // If there are no service options the record can not support variable subsetting
```
Comment (macrouch) [inline static/src/js/util/accessMethods/defaultConcatenation.js:8]: "the record can not support variable subsetting" Was this copied from another method? I don't see any variable subsetting in this method

File: static/src/js/util/accessMethods/supportsConcatenation.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Determine whether or not the provided UMM S record supports shapefile subsetting
+ * @param {Object} service UMM S record to parse
+ */
+export const supportsConcatenation = (service) => {
+  const { serviceOptions = {} } = service
+
+  // If there are no service options the record can not support variable subsetting
```
Comment (macrouch) [inline static/src/js/util/accessMethods/supportsConcatenation.js:8]: Same comment about variable subsetting

File: static/src/js/util/accessMethods/defaultConcatenation.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Determine whether or not the provided UMM S record supports shapefile subsetting
```
Comment (macrouch) [inline static/src/js/util/accessMethods/defaultConcatenation.js:2]: shapefile subsetting?

File: static/src/js/util/accessMethods/supportsConcatenation.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Determine whether or not the provided UMM S record supports shapefile subsetting
```
Comment (macrouch) [inline static/src/js/util/accessMethods/supportsConcatenation.js:2]: shapefile subsetting?

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js (hunk 8 above)
Comment (bnp26) [reply serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:698]: Could you check again and see that it's what you think it should be?

**Comment by bnp26** [comment]
> When checking the Enable Concatenation checkbox and refreshing the page, the form is not remembering the selected checkbox like it does with the other form fields. You need to be sure to put the value into the URL and populate it from the URL.

I added this as well as tests for it.

File: static/src/js/util/accessMethods/__tests__/defaultConcatenation.test.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        serviceOptions: {
+          aggregation: {}
+        }
+      })
+
+      expect(response).toBeFalsy()
+    })
+  })
+
+  describe('when concatenation is not supported', () => {
```
Comment (eudoroolivares2016) [inline static/src/js/util/accessMethods/__tests__/defaultConcatenation.test.js:34]: nit; isn't this the assertion that we want? Since even if concatenationDefault is false we are still "supporting" concatenation?
```suggestion
  describe('when concatenationDefault is set to false', () => {
```

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 14]
```diff
@@ -182,42 +187,38 @@ describe('buildAccessMethods', () => {
             ]
           }],
           variables: {
```
Comment (eudoroolivares2016) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:189]: I'm a bit confused why did the variables in the response need to go into the service associated variables?

**Review by eudoroolivares2016 — APPROVED** [review]

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js (hunk 8 above)
Comment (macrouch) [reply serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:698]:
```
      expect(response.getAll('concatenate')).toEqual([
        'true'
      ])
```

You aren't testing for either supportsConcatenation or enableConcatenateDownload, so the test name shouldn't say you are testing for them.

describe: describe the conditions for the test - 'when supportsConcatenation is true and enableConcatenateDownload is true'
test: describe the desired outcome of the test - 'constructs a payload with concatenate true'

All of the added tests in this file should follow that formula

**Review by macrouch — COMMENTED** [review]
Be sure to update https://wiki.earthdata.nasa.gov/display/EDSC/Earthdata+Search+URL+Parameters with the new URL parameter

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js (hunk 8 above)
Comment (bnp26) [reply serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:698]: I see. I'll change the tests to match that convention.

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js [hunk 15]
```diff
@@ -693,4 +693,115 @@ describe('constructOrderPayload', () => {
...
+      })
+
+      expect(response.getAll('concatenate')).toEqual([
+        'true'
+      ])
+    })
+  })
+
+  describe('when supportsConcatenation = false or enableConcatenateDownload = false', () => {
+    test('constructed payload does not containe concatenate', async () => {
```
Comment (macrouch) [inline serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:736]: containe

File: serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js [hunk 16]
```diff
@@ -693,4 +693,115 @@ describe('constructOrderPayload', () => {
...
+        granuleParams,
+        accessToken
+      })
+
+      expect(response.getAll('concatenate')).not.toEqual([
+        'true'
+      ])
+    })
+
+    test('constructed payload does not containe concatenate', async () => {
```
Comment (macrouch) [inline serverless/src/submitHarmonyOrder/__tests__/constructOrderPayload.test.js:772]: containe

**Review by rushgeo — APPROVED** [review]
Looks good to me. 

When running locally, the submission to Harmony fails because SQS isn't mocked in our dev environment.
After following the steps in the description, I followed these steps to submit the order to Harmony:

1. Make an `event.json` file that looks like this:
```json
{
    "Records": [
        {
            "body": {
                "accessToken": "",
                "id": 
            }
        }
    ]
}
```
2. Using TableTool or similar connected to your local Postgres DB, copy the `token` value from the last (newest) row in the `retrievals` table. Paste it into the empty quotes for `accessToken` in the JSON.
3. Copy the `id` value from the last row of the `retrieval_orders` table. Paste it into `event.json` as the value for `id`. Save the `event.json` file.
4. The value for `body` needs to be stringified, but everything outside of that needs to be valid JSON. Run these commands to make that conversion:
```sh
export body=$(jq -r '.Records[0].body' event.json)
jq -r '.Records[0].body = env.body' event.json | sed 's/\\n//g' > event_string.json
```
5. Run the local lambda for submitting the Harmony order with that JSON payload:
```sh
npm run invoke-local -- --function submitHarmonyOrder --path ./event_string.json
```
6. Confirm the order was submitted at https://harmony.uat.earthdata.nasa.gov/workflow-ui
